scipy/sparse/linalg/tests/test_special_sparse_arrays.py

```diff
@@ -0,0 +1,113 @@
+                [
+                    [-2, 1, 0, 0, 0, 0],
+                    [1, -2, 1, 0, 0, 0],
+                    [0, 1, -2, 1, 0, 0],
+                    [0, 0, 1, -2, 1, 0],
+                    [0, 0, 0, 1, -2, 1],
+                    [0, 0, 0, 0, 1, -2],
+                ]
+            )
+        else:
+            a = np.array(
+                [
+                    [-2, 1, 0, 0, 0, 1],
+                    [1, -2, 1, 0, 0, 0],
+                    [0, 1, -2, 1, 0, 0],
+                    [0, 0, 1, -2, 1, 0],
+                    [0, 0, 0, 1, -2, 1],
+                    [1, 0, 0, 0, 1, -2],
+                ]
+            )
+        assert_array_equal(a, lapa)
+
+
+    def test_1d_with_graph_laplacian(self):
+        n = 6
+        G = diags(np.ones(n - 1), 1, format="dia")
+        Lf = csgraph.laplacian(G, symmetrized=True, form="function")
+        La = csgraph.laplacian(G, symmetrized=True, form="array")
+        grid_shape = (n,)
+        bc = "Neumann"
+        lap = LaplacianNd(grid_shape, boundary_conditions=bc)
+        assert_array_equal(lap(np.eye(n)), -Lf(np.eye(n)))
+        assert_array_equal(lap.toarray(), -La.toarray())
+        # https://github.com/numpy/numpy/issues/24351
+        assert_array_equal(lap.tosparse().toarray(), -La.toarray())
+
+
+    @pytest.mark.parametrize("grid_shape", [(6, ), (2, 3), (2, 3, 4)])
+    @pytest.mark.parametrize("bc", ['Neumann', 'Dirichlet', 'Periodic'])
+    def test_eigenvalues(self, grid_shape, bc):
+        lap = LaplacianNd(grid_shape, boundary_conditions=bc)
+        eigenvalues = lap.eigenvalues
+        # Need to retype the `int8` array into `float` for full accuracy 
+        l = lap.toarray().astype(np.float64)
+        eigvals = np.sort(eigh(l, eigvals_only=True))
+        dtype = eigenvalues.dtype
+        n = np.prod(grid_shape)
+        atol = n * n * max(np.finfo(dtype).eps, np.finfo(np.double).eps)
+        assert_allclose(eigenvalues, eigvals, atol=atol)
+
+    @pytest.mark.parametrize("grid_shape", [(6, ), (2, 3), (2, 3, 4)])
+    @pytest.mark.parametrize("bc", ['Neumann', 'Dirichlet', 'Periodic'])
+    def test_toarray_tosparse_consistency(self, grid_shape, bc):
+        lap = LaplacianNd(grid_shape, boundary_conditions=bc)
+        n = np.prod(grid_shape)
+        assert_array_equal(lap.toarray(), lap(np.eye(n)))
+        assert_array_equal(lap.tosparse().toarray(), lap.toarray())
+
+
+    @pytest.mark.parametrize("dtype", ALLDTYPES)
+    @pytest.mark.parametrize("grid_shape", [(6, ), (2, 3), (2, 3, 4)])
+    @pytest.mark.parametrize("bc", ['Neumann', 'Dirichlet', 'Periodic'])
+    def test_input_output_dtype_shape_consistency(self, grid_shape, bc, dtype):
+        lap = LaplacianNd(grid_shape, boundary_conditions=bc)
+        n = np.prod(grid_shape)
+        x0 = np.arange(n)
+        x1 = x0.reshape((-1, 1))
+        x2 = np.arange(2 * n).reshape((n, 2))
+        input_set = [x0, x1, x2]
+        for x in input_set:
+            y = lap(x.astype(dtype))
```

Comment:
`lap` is an instance of a LaplacianND and hence don't know what it means here.

Reply:
it runs matmat

Reply:
That should me explicit. I'll fix it

Reply:
`LinearOperator.__call__(self, x)` is defined as `self * x`

Reply:
with the goal to test that matmat(x) preserves the dtype of x

Reply:
I us this same call in the docsting multiple times. What do you mean "explicit"?

Reply:
I switched to using sparse arrays and hence getting the error due to 

https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/_interface.py#L331-L332

Hence we need to separate the 1D case.

Reply:
LinearOperator just calls matvec in 1D case. What's wrong with `lap(x)`? It does look and feel like a function call and that is what this is. Why do you want to call matmat directly? I do not believe this is a standard practice for LienarOpearators.